### PR TITLE
Onboard IaaS security groups

### PR DESCRIPTION
**Pull request by GokceGK**

### CODE REVIEW

File: stackit/internal/services/iaas/securitygroup/resource.go [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+
+type Model struct {
+	Id              types.String `tfsdk:"id"` // needed by TF
+	ProjectId       types.String `tfsdk:"project_id"`
+	SecurityGroupId types.String `tfsdk:"security_group_id"`
+	Name            types.String `tfsdk:"name"`
+	Description     types.String `tfsdk:"description"`
+	Labels          types.Map    `tfsdk:"labels"`
+	Stateful        types.Bool   `tfsdk:"stateful"`
+	Rules           types.List   `tfsdk:"rules"`
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/resource.go:53]: Security Group rules will be a separate resource, so it shouldn't be included here

File: stackit/internal/services/iaas/securitygroup/resource.go [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+				Validators: []validator.String{
+					validate.UUID(),
+					validate.NoSeparator(),
+				},
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of the security group.",
+				Required:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/resource.go:216]: this is not needed when a field is `Required`

File: stackit/internal/services/iaas/securitygroup/resource.go [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+					stringvalidator.LengthAtMost(127),
+				},
+			},
+			"labels": schema.MapAttribute{
+				Description: "Labels are key-value string pairs which can be attached to a resource container",
+				ElementType: types.StringType,
+				Optional:    true,
+			},
+			"stateful": schema.BoolAttribute{
+				Description: "Shows if a security group is stateful or stateless. There can only be one security groups per network interface/server.",
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/resource.go:244]:
```suggestion
				Description: "Shows if a security group is stateful or stateless. There can only be one security group per network interface/server.",
```

File: stackit/internal/services/iaas/securitygroup/resource.go [hunk 4]
```diff
@@ -0,0 +1,708 @@
...
+				Optional:    true,
+				Computed:    true,
+			},
+			"rules": schema.ListNestedAttribute{
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/resource.go:248]: remove this field from the schema, separate resource

File: stackit/internal/services/iaas/securitygroup/resource.go [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+
+	idParts := []string{
+		model.ProjectId.ValueString(),
+		securityGroupId,
+	}
+	model.Id = types.StringValue(
+		strings.Join(idParts, core.Separator),
+	)
+
+	var labels basetypes.MapValue
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/resource.go:577]: refactor this to match what is implemented in https://github.com/stackitcloud/terraform-provider-stackit/blob/8894383c2c4e4f0036d061f594c0c40200de0aff/stackit/internal/services/iaas/volume/resource.go#L522

File: stackit/internal/services/iaas/securitygroup/resource.go [hunk 6]
```diff
@@ -0,0 +1,708 @@
...
+
+	err := mapRules(securityGroupResp, model)
+	if err != nil {
+		return fmt.Errorf("map node_pools: %w", err)
+	}
+
+	return nil
+}
+
+func mapRules(securityGroup *iaasalpha.SecurityGroup, model *Model) error {
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/resource.go:602]: move to separate resource

File: stackit/internal/services/iaas/securitygroup/datasource.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+			},
+			"security_group_id": schema.StringAttribute{
+				Description: "The security group ID.",
+				Required:    true,
+				Validators: []validator.String{
+					validate.UUID(),
+					validate.NoSeparator(),
+				},
+			},
+			"server_id": schema.StringAttribute{
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/datasource.go:115]: remove this, there is no `server_id` field

File: stackit/internal/services/iaas/securitygroup/datasource.go [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+			"labels": schema.MapAttribute{
+				Description: "Labels are key-value string pairs which can be attached to a resource container",
+				ElementType: types.StringType,
+				Computed:    true,
+			},
+			"stateful": schema.BoolAttribute{
+				Description: "Shows if a security group is stateful or stateless. There can only be one security groups per network interface/server.",
+				Computed:    true,
+			},
+			"rules": schema.ListNestedAttribute{
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/datasource.go:140]: move to separate datasource

**Comment by vicentepinto98** [comment]
Merge with feature/iaas-release branch

File: examples/data-sources/stackit_security_group/resource.tf [hunk 9]
```diff
@@ -0,0 +1,4 @@
+resource "stackit_security_group" "example" {
```
Comment (vicentepinto98) [inline examples/data-sources/stackit_security_group/resource.tf:1]: this file should be called `datasource.tf`

Comment (GokceGK) [reply examples/data-sources/stackit_security_group/resource.tf]: renamed the file

File: stackit/internal/services/iaas/securitygroup/datasource.go (hunk 7 above)
Comment (GokceGK) [reply stackit/internal/services/iaas/securitygroup/datasource.go:115]: removed it from data source and docs

File: stackit/internal/services/iaas/securitygroup/resource.go (hunk 2 above)
Comment (GokceGK) [reply stackit/internal/services/iaas/securitygroup/resource.go:216]: thanks, removed it

File: stackit/internal/services/iaas/securitygroup/resource.go (hunk 5 above)
Comment (GokceGK) [reply stackit/internal/services/iaas/securitygroup/resource.go]: refactored the code block

File: stackit/internal/services/iaas/securitygroup/resource.go (hunk 1 above)
Comment (GokceGK) [reply stackit/internal/services/iaas/securitygroup/resource.go]: removed rules from schema

File: stackit/internal/services/iaas/securitygroup/resource.go (hunk 4 above)
Comment (GokceGK) [reply stackit/internal/services/iaas/securitygroup/resource.go]: removed rules from the schema. I will add rules in a separate branch

File: stackit/internal/services/iaas/securitygroup/resource.go (hunk 6 above)
Comment (GokceGK) [reply stackit/internal/services/iaas/securitygroup/resource.go]: removed rules from the schema. I will add rules in a separate branch

File: stackit/internal/services/iaas/securitygroup/datasource.go (hunk 8 above)
Comment (GokceGK) [reply stackit/internal/services/iaas/securitygroup/datasource.go]: removed rules from the schema. I will add rules in a separate branch

File: stackit/internal/services/iaas/securitygroup/resource.go [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+					stringvalidator.LengthAtMost(127),
+				},
+			},
+			"labels": schema.MapAttribute{
+				Description: "Labels are key-value string pairs which can be attached to a resource container",
+				ElementType: types.StringType,
+				Optional:    true,
+			},
+			"stateful": schema.BoolAttribute{
+				Description: "Shows if a security group is stateful or stateless. There can only be one security group per network interface/server.",
```
Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/resource.go:175]:
```suggestion
				Description: "Configures if a security group is stateful or stateless.",
```

About "There can only be one security group per network interface/server.", I don't fully get it, can you clarify with the team what it means?

Is that sentence only relevant for stateful security groups?

Comment (vicentepinto98) [inline stackit/internal/services/iaas/securitygroup/resource.go:175]: This field should also have `RequiresReplace` and `UseStateForUnknown`

Comment (GokceGK) [reply stackit/internal/services/iaas/securitygroup/resource.go:175]: Description has been changed in the meantime in API docs. It is now: "Shows if a security group is stateful or stateless. You can only have one type of security groups per network interface/server."

The limitation is not one security group per network interface. It is one type of security group. So in a network interface, there can be either stateful or stateless security groups. 

Would that be more understandable if I changed to : 

"Configures if a security group is stateful or stateless. There can only be one type of security groups per network interface/server."

Comment (vicentepinto98) [reply stackit/internal/services/iaas/securitygroup/resource.go:176]: Yes, that looks good

**Review by vicentepinto98 — APPROVED** [review]

**Pull request merged**